### PR TITLE
Relative Line Number support 

### PR DESCRIPTION
**Pull request by alfanick**
Adds support for vim-style relative line numbers.

Useful for vim motion plugins, replacing extension "Relative Line Numbers" by extr0py.
- word-wrapping aware
- adds `editor.relativeLineNumbers` option (changeable without restart), default to false
- current line is absolute, others are relative (behaviour of hybrid relative numbers in vim)

### CODE REVIEW

**Comment by alfanick** [comment]
I've made some more aggresive optimizations on [alfanick/vscode/optimized-relative-numbers](https://github.com/alfanick/vscode/tree/optimized-relative-numbers) – I'm assuming that there is no need for rendering line numbers on modelLinesInserted/modelLinesDeleted as cursorPositionChanged is always called with these events, also no need for rendering with every modelLineChanged unless wordWrapping is enabled (this is only case when line can break and change numbering). 

I'm not really expert on VSCode flow, so these optimizations actually may break smth, but they work for me – the pull request does not brake anything, where branch optimized-relative-numbers does not brake anything either, but needs thorough code review.

File: src/vs/editor/browser/viewParts/lineNumbers/lineNumbers.ts [hunk 1]
```diff
@@ -57,7 +61,15 @@ export class LineNumbersOverlay extends DynamicViewOverlay {
...
 	}
 	public onCursorPositionChanged(e:editorCommon.IViewCursorPositionChangedEvent): boolean {
-		return false;
+		let modelPosition = this._context.model.convertViewPositionToModelPosition(e.position.lineNumber, e.position.column);
+
+		if (!this._relativeLineNumbers || this._currentLineNumber === modelPosition.lineNumber) {
+			return false;
+		}
+
+		this._currentLineNumber = modelPosition.lineNumber;
```
Comment (alexandrudima) [inline src/vs/editor/browser/viewParts/lineNumbers/lineNumbers.ts]: In the case where `this._relativeLineNumbers` is `false`, `this._currentLineNumber` should still be assigned such that `this._currentLineNumber` always has a good value.

It is possible that the `relativeLineNumbers` option changes at runtime and in that case we'd want `this._currentLineNumber` to be up-to-date.

File: src/vs/editor/browser/viewParts/lineNumbers/lineNumbers.ts [hunk 2]
```diff
@@ -103,22 +118,31 @@ export class LineNumbersOverlay extends DynamicViewOverlay {
...
+		let commonLeft = '<div class="' + ClassNames.LINE_NUMBERS + lineHeightClassName + '" style="text-align:left;left:' + this._lineNumbersLeft.toString() + 'px;width:' + this._lineNumbersWidth.toString() + 'px;height:' + lineHeight + 'px;">';
 
 		let output: string[] = [];
 		for (let lineNumber = visibleStartLineNumber; lineNumber <= visibleEndLineNumber; lineNumber++) {
 			let lineIndex = lineNumber - visibleStartLineNumber;
+			let position = this._context.model.convertViewPositionToModelPosition(lineNumber, 1);
+			let relativeLineNumber = position.lineNumber;
+
+			if (position.column === 1) {
+				if (this._relativeLineNumbers) {
```
Comment (alexandrudima) [inline src/vs/editor/browser/viewParts/lineNumbers/lineNumbers.ts]: The test `if (this._relativeLineNumbers)` should be the first thing done here before doing any position conversions. All the new rendering code should go in this `if`. All the old rendering code should be maintained in an `else` branch. As it looks right now, regular line number rendering appears to be broken.

File: .gitignore [hunk 3]
```diff
@@ -3,6 +3,7 @@ npm-debug.log
 Thumbs.db
 node_modules/
 .build/
+.vscode/.browse.VC*
```
Comment (alexandrudima) [inline .gitignore]: Please remove this one, you can use `.git/info/exclude` for it

**Review by alexdima — COMMENTED** [review]
Thank you!

The change gets correctly the adding an option code, and when to update/re-render the line numbers.

However, it appears to me that it breaks the old behaviour of non-relative line numbers completely.

Please address these comments, and resubmit the PR, as it appears many unnecessary commits have made their way into this PR.

Looking forward to your changes! 👍

**Comment by alexandrudima** [comment]
@alfanick Please let me know if this is something you want to look into, otherwise I will pick it up, as it is something we want to get done in Sept (i.e. by Friday the latest). Then I need to expose extension host API for it.

**Comment by alexandrudima** [comment]
I've implemented this in 13bbe32660baa92557d83e831e1b8a8b5cf4c8c0

**Pull request closed without merging**